### PR TITLE
Handle unhandled errors and improve error handling

### PR DESCRIPTION
**Pull request by romshark**
- Replace panics with returned errors.
- Replace panics with fatal logs.
- Handle unhandled errors.
- Improve error reporting according to conventions.

### CODE REVIEW

File: goudpscan/scanner.go [hunk 1]
```diff
@@ -96,27 +97,23 @@ func BreakUPIP(segments []string) []string {
...
 		}
 
 		for i := start; i <= end; i++ {
-			tails = append(tails,
-				BreakUPIP(
-					append(
-						[]string{strconv.Itoa(i)}, segments[idx+1:]...,
-					),
-				)...,
+			r, err := BreakUPIP(
```
Comment (KernelPryanic) [inline goudpscan/scanner.go:108]: Why to allocate another variable?

File: goudpscan/scanner.go [hunk 2]
```diff
@@ -96,27 +97,23 @@ func BreakUPIP(segments []string) []string {
...
+				append(
+					[]string{strconv.Itoa(i)}, segments[idx+1:]...,
+				),
 			)
+			if err != nil {
+				return nil, fmt.Errorf("breaking UPIP: %w", err)
```
Comment (KernelPryanic) [inline goudpscan/scanner.go:114]: Please use something like "failed to break up IP"

File: goudpscan/scanner.go [hunk 3]
```diff
@@ -127,26 +124,31 @@ func BreakUPIP(segments []string) []string {
...
 		}
-	} else {
-		subnets = BreakUPIP(segments)
 	}
 
-	return subnets
+	subnets, err := BreakUPIP(segments)
+	if err != nil {
+		return nil, fmt.Errorf("breaking UPIP: %w", err)
+	}
```
Comment (KernelPryanic) [inline goudpscan/scanner.go]: The logic is broken here, we should only parse the whole `segments` array if there's no network definition "/"

File: goudpscan/scanner.go [hunk 4]
```diff
@@ -158,19 +160,13 @@ func BreakUPPort(portRange string) []uint16 {
 
 			start64, err := strconv.ParseUint(_start, 10, 16)
 			if err != nil {
-				panic(fmt.Errorf(
-					"Error in breaking up port ranges: %s",
-					err,
-				))
+				return nil, fmt.Errorf("parsing start uint: %w", err)
```
Comment (KernelPryanic) [inline goudpscan/scanner.go:164]: Please use something like "failed to parse port range start"

File: goudpscan/scanner.go [hunk 5]
```diff
@@ -158,19 +160,13 @@ func BreakUPPort(portRange string) []uint16 {
...
 			}
 			start := uint16(start64)
 
 			end64, err := strconv.ParseUint(_end, 10, 16)
 			if err != nil {
-				panic(fmt.Errorf(
-					"Error in breaking up port ranges: %s",
-					err,
-				))
+				return nil, fmt.Errorf("parsing end uint: %w", err)
```
Comment (KernelPryanic) [inline goudpscan/scanner.go:170]: Please use something like "failed to parse port range end"

File: goudpscan/scanner.go [hunk 6]
```diff
@@ -180,17 +176,14 @@ func BreakUPPort(portRange string) []uint16 {
 		} else {
 			port64, err := strconv.ParseUint(s, 10, 16)
 			if err != nil {
-				panic(fmt.Errorf(
-					"Error in breaking up port ranges: %s",
-					err,
-				))
+				return nil, fmt.Errorf("breaking up port ranges: %w", err)
```
Comment (KernelPryanic) [inline goudpscan/scanner.go:180]: Please use something like "failed to parse port"

File: goudpscan/scanner.go [hunk 7]
```diff
@@ -302,26 +298,25 @@ func waitResponse(conn net.Conn,
...
-func SniffICMP(ch chan bool, wg *sync.WaitGroup) {
+func SniffICMP(ch chan bool, wg *sync.WaitGroup) error {
 	defer wg.Done()
 	conn, err := icmp.ListenPacket("ip4:icmp", "0.0.0.0")
 	if err != nil {
-		panic(fmt.Errorf(
-			"Can't open icmp socket to listen on: %s",
-			err,
-		))
+		return fmt.Errorf("opening icmp socket: %w", err)
```
Comment (KernelPryanic) [inline goudpscan/scanner.go:320]: "can't open ICMP socket" or "failed to open ICMP socket"

File: goudpscan/scanner.go [hunk 8]
```diff
@@ -344,17 +339,11 @@ func SniffICMP(ch chan bool, wg *sync.WaitGroup) {
 				netOpError.Err.Error() == "use of closed network connection" {
 				break
 			}
-			panic(fmt.Errorf(
-				"Something wrong with reading from ICMP socket: %s",
-				err,
-			))
+			return fmt.Errorf("reading from ICMP socket: %w", err)
```
Comment (KernelPryanic) [inline goudpscan/scanner.go:343]: ditto

File: goudpscan/scanner.go [hunk 9]
```diff
@@ -344,17 +339,11 @@ func SniffICMP(ch chan bool, wg *sync.WaitGroup) {
...
 		}
 		rm, err := icmp.ParseMessage(iana.ProtocolICMP, rb[:n])
 		if err != nil {
-			panic(fmt.Errorf(
-				"Something wrong with parsing ICMP response: %s",
-				err,
-			))
+			return fmt.Errorf("parsing ICMP response: %w", err)
```
Comment (KernelPryanic) [inline goudpscan/scanner.go:347]: ditto

File: goudpscan/scanner.go [hunk 10]
```diff
@@ -369,20 +358,29 @@ func SniffICMP(ch chan bool, wg *sync.WaitGroup) {
...
-func (s scanner) Scan(ch chan bool) map[string]string {
+func (s scanner) Scan(errLog *log.Logger, ch chan bool) (map[string]string, error) {
 	throttle := make(chan int, s.opts.maxConcurrency)
 	subnets := []string{}
 
 	for _, host := range s.hosts {
-		subnets = append(subnets, ParseSubnet(host)...)
+		r, err := ParseSubnet(host)
+		if err != nil {
+			return nil, fmt.Errorf("parsing subnet: %w", err)
```
Comment (KernelPryanic) [inline goudpscan/scanner.go:372]: ditto

File: goudpscan/scanner.go [hunk 11]
```diff
@@ -369,20 +358,29 @@ func SniffICMP(ch chan bool, wg *sync.WaitGroup) {
...
+		subnets = append(subnets, r...)
 	}
 
 	ports := []uint16{}
 
 	for _, port := range s.ports {
-		ports = append(ports, BreakUPPort(port)...)
+		r, err := BreakUPPort(port)
+		if err != nil {
+			return nil, fmt.Errorf("breaking up port: %w", err)
```
Comment (KernelPryanic) [inline goudpscan/scanner.go:382]: ditto

File: goudpscan/scanner.go [hunk 12]
```diff
@@ -392,16 +390,14 @@ func (s scanner) Scan(ch chan bool) map[string]string {
 			defer wgSubnets.Done()
 			ips, err := Hosts(subnet)
 			if err != nil {
-				panic(fmt.Errorf(
-					"Error in ip generating: %s",
-					err,
-				))
+				errLog.Printf("generating ip: %s", err)
```
Comment (KernelPryanic) [inline goudpscan/scanner.go:394]: ditto

File: main.go [hunk 13]
```diff
@@ -122,11 +123,14 @@ func MergeAsync(left []string, right []string, resultChannel chan []string) {
...
 
-func FormPayload(payloadData map[string][]string) map[uint16][]string {
+func FormPayload(payloadData map[string][]string) (map[uint16][]string, error) {
 	payload := map[uint16][]string{}
 
 	for k, v := range payloadData {
-		ports := goudpscan.BreakUPPort(k)
+		ports, err := goudpscan.BreakUPPort(k)
+		if err != nil {
+			return nil, fmt.Errorf("breaking up port: %w", err)
```
Comment (KernelPryanic) [inline main.go:132]: ditto

File: main.go [hunk 14]
```diff
@@ -151,32 +157,37 @@ func main() {
 		payloadFile, err = os.ReadFile(*payloads)
 	}
 	if err != nil {
-		panic(fmt.Errorf(
-			"Failed reading file with payloads: %s",
-			err,
-		))
+		errl.Fatalf("reading file with payloads: %s", err)
```
Comment (KernelPryanic) [inline main.go:160]: ditto

File: main.go [hunk 15]
```diff
@@ -151,32 +157,37 @@ func main() {
...
-		os.Exit(0)
+		return
 	}
 	payloadData := make(map[string][]string)
 	if err = yaml.Unmarshal(payloadFile, &payloadData); err != nil {
-		panic(fmt.Errorf(
-			"Failed parsing file with payloads: %s",
-			err,
-		))
+		errl.Fatalf("parsing payloads file: %s", err)
```
Comment (KernelPryanic) [inline main.go:168]: ditto

File: main.go [hunk 16]
```diff
@@ -151,32 +157,37 @@ func main() {
...
 	}
 
 	var wg sync.WaitGroup
 	if !*fast {
 		wg.Add(1)
-		go goudpscan.SniffICMP(ch, &wg)
+		go func() {
+			if err := goudpscan.SniffICMP(ch, &wg); err != nil {
+				errl.Printf("sniffing ICMP: %s", err)
```
Comment (KernelPryanic) [inline main.go:176]: ditto

File: main.go [hunk 17]
```diff
@@ -151,32 +157,37 @@ func main() {
...
+			}
+		}()
 	}
-	sc := goudpscan.New(*hosts, *ports, FormPayload(payloadData), &opts)
+	pl, err := FormPayload(payloadData)
+	if err != nil {
+		errl.Fatalf("forming payload: %s", err)
```
Comment (KernelPryanic) [inline main.go:182]: ditto

File: main.go [hunk 18]
```diff
@@ -151,32 +157,37 @@ func main() {
...
+	}
+	sc := goudpscan.New(*hosts, *ports, pl, &opts)
 
 	start := time.Now()
-	result := sc.Scan(ch)
+	result, err := sc.Scan(errl, ch)
+	if err != nil {
+		errl.Fatalf("scanning: %s", err)
```
Comment (KernelPryanic) [inline main.go:189]: ditto

**Review by KernelPryanic — CHANGES_REQUESTED** [review]

File: goudpscan/scanner.go (hunk 2 above)
Comment (romshark) [reply goudpscan/scanner.go:114]: "failed to" doesn't make make sense in an error message. An error message already indicates failure. What's interesting to know is what you were doing when something went wrong.

Comment (KernelPryanic) [reply goudpscan/scanner.go:114]: I see, but there was a failure already and the reason why it happened is following the statement. From this message you also get the same information of what was happening)

File: goudpscan/scanner.go (hunk 3 above)
Comment (romshark) [reply goudpscan/scanner.go]: Fixed ✅

File: goudpscan/scanner.go (hunk 2 above)
Comment (romshark) [reply goudpscan/scanner.go:114]: Whenever you get an `error` in return the failure already happened. There's no need to explain that explicitly.

```
failed to x: failed to y: failed to z: p was wrong
```
vs
```
doing x: doing y: doing z: p was wrong
```
You make your error messages less verbose.

What happened is described by the source error. In the chain of errors you just describe what you were doing when it happened.

Comment (KernelPryanic) [reply goudpscan/scanner.go:114]: Okay, after the discussion we've agreed on the simple form of the error messages, as it's generally more suitable. Example:
```
read config: open file: file doesn't exist
```

**Pull request closed without merging**